backends/zypp/pk-backend-zypp.cpp

```diff
@@ -1722,18 +1723,15 @@ zypp_emit_filtered_packages_in_list (PkBackendJob *job, PkBitfield filters, cons
 		}
 
 		for (sat_it_t it = available.begin (); it != available.end (); ++it) {
-			bool match = false;
-
-			for (sat_it_t i = installed.begin (); !match && i != installed.end (); i++) {
-				match = it->sameNVRA (*i) &&
+			for (sat_it_t it2 = installed.begin (); it2 != installed.end (); it2++) {
+				if (it->sameNVRA (*it2) &&
 					!(!isKind<SrcPackage>(*it) ^
-					  !isKind<SrcPackage>(*i));
-			}
-
-			if (match) {
-				zypp_backend_package (job, PK_INFO_ENUM_INSTALLED, *it,
-						      make<ResObject>(*it)->summary().c_str(), true);
-				installed.erase (find (installed.begin (), installed.end(), *it));
+					  !isKind<SrcPackage>(*it2))) {
+					zypp_backend_package (job, PK_INFO_ENUM_INSTALLED, *it,
+							      make<ResObject>(*it)->summary().c_str(), true);
+					installed.erase (it2);
+					break;
+				}
```

Comment:
The old code here broke my brain. The `match` variable gets overwritten every inner loop run, so only the last loop result counts. Maybe it should be `match |=` instead? This could well explain why the crash only happens once in a while.

According to [documentation](https://doc.opensuse.org/projects/libzypp/11.2/classzypp_1_1sat_1_1Solvable.html), this checks that the packages have the same name, version, release and architecture and are they source packages or not:

```c++
match = it->sameNVRA(*i) && !(!isKind<SrcPackage>(*it) ^ !isKind<SrcPackage>(*i));
```

Which AFAICT is logically identical to this:

```c++
match = it->sameNVRA(*i) && isKind<SrcPackage>(*it) == isKind<SrcPackage>(*i);
```

I'm wondering if the `&& isKind<SrcPackage>(*it) == isKind<SrcPackage>(*i)` could be simplified to `&& it->kind() == i->kind()` as it would make sense to check if name, version, release and architecture and package-kind matches... Maybe there's a reason to only check source-or-not, I didn't dig any deeper.

Anyway, the new code makes a lot more sense!

Reply:
> The old code here broke my brain. The match variable gets overwritten every inner loop run, so only the last loop result counts. Maybe it should be match |= instead? This could well explain why the crash only happens once in a while.

To note that the loop condition was "!match && i != installed.end ()" instead of the common "i != installed.end()" so that should have worked good. But maybe was not that good idea to introduce a separate variable instead of just "break;"ing out of the loop etc.

> I'm wondering if the && isKind<SrcPackage>(*it) == isKind<SrcPackage>(*i) could be simplified to && it->kind() == i->kind() as it would make sense to check if name, version, release and architecture and package-kind matches... Maybe there's a reason to only check source-or-not, I didn't dig any deeper.

Yea, was also thinking that it looks like a suspicious comparison, but apparently works. There are couple other existing cases instances of such so maybe just keep it for now. I'd suppose the comparison here was just copy pasted.

In general there are suspicious things going on with the zypper backend here, and also on what I've been occasionally checking libzypp side. I don't have the highest trust on the these pieces of code.